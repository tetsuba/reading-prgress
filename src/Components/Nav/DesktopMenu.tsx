import { NavLink, useNavigate } from 'react-router-dom'
import Button from '../Button/Button'
import { useDispatch } from 'react-redux'
import SignOutButton from './SignOutButton'

function getClasses({ isActive }: { isActive: boolean }) {
    const defaultStyles = 'block px-4 py-2 text-sm round-xl'
    return isActive
        ? `bg-gray-100 text-gray-300 ${defaultStyles}`
        : `text-gray-700 hover:bg-gray-200 ${defaultStyles}`
}

export default function DesktopMenu() {
    const navigate = useNavigate()
    const dispatch = useDispatch()

    return (
        <div
            data-testid="user-menu"
            className="group/menu relative hidden items-center text-end md:flex"
        >
            <Button
                data-testid="user-menu-button"
                icon="user"
                template="icon-user"
            />
            <div
                data-testid="user-menu-open"
                className="absolute right-0 top-6 mt-2 hidden w-48 rounded-md border-2 border-gray-300 bg-white py-2 text-left shadow-xl group-hover/menu:block"
                role="menu"
            >
                <NavLink className={getClasses} to="/profile">
                    Your Profile
                </NavLink>
                <NavLink className={getClasses} to="/settings">
                    Settings
                </NavLink>
                <SignOutButton template="nav-menu-desktop" />
            </div>
        </div>
    )
}
