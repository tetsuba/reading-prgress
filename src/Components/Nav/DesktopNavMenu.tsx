import { NavLink, useNavigate } from 'react-router-dom'
import Button from '../Button/Button'
import { resetUserToInitialState } from '../../store/user/userSlice'
import ls from '../../lib/localStorage'
import { useDispatch } from 'react-redux'
import Svg from '../Svg/Svg'

function getClasses({ isActive }: { isActive: boolean }) {
    const defaultStyles = 'block px-4 py-2 text-sm round-xl'
    return isActive
        ? `bg-gray-100 text-gray-300 ${defaultStyles}`
        : `text-gray-700 hover:bg-gray-200 ${defaultStyles}`
}

export default function DesktopNavMenu() {
    const navigate = useNavigate()
    const dispatch = useDispatch()

    return (
        <div
            data-testid="user-menu"
            className="group/menu hidden pt-4 text-end md:block"
        >
            <Button data-testid="user-menu-button" template="svgUser">
                <Svg icon="user" />
            </Button>
            <div
                data-testid="user-menu-open"
                className="absolute right-8 top-12 mt-2 hidden w-48 rounded-md border-2 border-gray-300 bg-white py-2 text-left shadow-xl group-hover/menu:block"
                role="menu"
            >
                <NavLink className={getClasses} to="/profile">
                    Your Profile
                </NavLink>
                <NavLink className={getClasses} to="/settings">
                    Settings
                </NavLink>
                <Button
                    type="button"
                    template="desktopNavMenuSignOut"
                    onClick={() => {
                        dispatch(resetUserToInitialState())
                        ls.remove()
                        navigate('/')
                    }}
                >
                    Sign out
                </Button>
            </div>
        </div>
    )
}
