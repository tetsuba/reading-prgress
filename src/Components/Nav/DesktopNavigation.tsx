import { NavLink } from 'react-router-dom'
import { useSelector } from 'react-redux'
import { userTokenSelector } from '../../store/user/userSelectors'
import Display from '../Dispay/Display'

export default function DesktopNavigation() {
    const userToken = useSelector(userTokenSelector)
    const DEFAULT_CLASSES = 'text-sm font-medium rounded-md px-3 py-2'
    const NOT_ACTIVE = `${DEFAULT_CLASSES} text-gray-300 hover:bg-gray-700 hover:text-white`
    const ACTIVE = `${DEFAULT_CLASSES}  bg-gray-900 text-white`
    function getClasses({ isActive }: { isActive: boolean }) {
        return isActive ? ACTIVE : NOT_ACTIVE
    }

    return (
        <div className="ml-10 hidden items-baseline space-x-4 md:flex">
            <Display value={!!userToken}>
                <>
                    <NavLink className={getClasses} to="/dashboard">
                        Dashboard
                    </NavLink>
                    <NavLink className={getClasses} to="books">
                        Books
                    </NavLink>
                    <NavLink className={getClasses} to="reading">
                        Reading
                    </NavLink>
                    <NavLink className={getClasses} to="sight-words">
                        Sight Words
                    </NavLink>
                </>
            </Display>
            <Display value={!userToken}>
                <>
                    <NavLink className={getClasses} to="/">
                        Home
                    </NavLink>
                    <NavLink className={getClasses} to="register">
                        Register
                    </NavLink>
                </>
            </Display>
        </div>
    )
}
