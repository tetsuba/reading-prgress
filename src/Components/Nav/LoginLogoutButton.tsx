import Button from "../Button/Button";
import {useState} from "react";
import {useNavigate} from "react-router-dom";
import Modal from "../Modal/Modal";
import Login from "../../Views/Login";
import ls from "../../lib/localStorage";

const classNames = 'inline-block rounded-lg px-3 py-1.5 text-sm font-semibold leading-6 text-gray-900 shadow-sm ring-1 ring-gray-900/10 hover:ring-gray-900/20'
export default function LoginLogoutButton() {
    const [showLogin, setShowLogin] = useState(false)
    const navigate = useNavigate();
    const token = ls.get()

    return !token
        ? (
            <>
                <Button
                    clickHandler={() => setShowLogin(true)}
                    className={classNames}>
                    Log in
                </Button>
                {
                    showLogin && (
                        showLogin && (
                            <Modal><Login setShowLogin={setShowLogin} /></Modal>
                        )
                    )
                }
            </>
        ) : (
            <Button
                clickHandler={() => {
                    ls.remove()
                    navigate('/')
                }}
                className={classNames}>
                Log out
            </Button>
        )
}


