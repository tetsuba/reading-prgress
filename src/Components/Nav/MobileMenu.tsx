import { useLocation, useNavigate } from 'react-router-dom'
import Button from '../Button/Button'
import { resetUserToInitialState } from '../../store/user/userSlice'
import ls from '../../lib/localStorage'
import { useDispatch } from 'react-redux'
import { useState } from 'react'

type PropTypes = {
    closeMenu: () => void
}
export default function MobileMenu(props: PropTypes) {
    const dispatch = useDispatch()
    const navigate = useNavigate()
    const location = useLocation()
    const [links] = useState([
        { name: 'Dashboard', path: 'dashboard' },
        { name: 'Books', path: 'books' },
        { name: 'Reading', path: 'reading' },
        { name: 'Sight Words', path: 'sight-words' },
        { name: 'Profile', path: 'profile' },
        { name: 'Settings', path: 'settings' }
    ])

    function isActive(locationPath: string, linkPath: string) {
        return locationPath.includes(linkPath)
            ? 'text-gray-300 bg-gray-100'
            : 'text-gray-800 bg-white'
    }

    function onClick(path: string) {
        navigate(`/${path}`)
        props.closeMenu()
    }

    return (
        <div data-testid="mobile-nav-menu" className="flex flex-col pt-10">
            <Button
                template="icon"
                className="absolute right-1 top-1.5 ml-auto p-1.5"
                dataTestid="modal-close"
                type="button"
                svg="close"
                clickHandler={() => props.closeMenu()}
            />
            {links.map((link, i) => (
                <Button
                    dataTestid="mobile-menu-link"
                    key={`mobile-menu-${i}`}
                    type="button"
                    template="none"
                    clickHandler={() => onClick(link.path)}
                    className={`py-4 text-center text-2xl ${isActive(
                        location.pathname,
                        link.path
                    )}`}
                >
                    {link.name}
                </Button>
            ))}

            <Button
                dataTestid="sign-out-mobile-menu"
                type="button"
                template="none"
                clickHandler={() => {
                    dispatch(resetUserToInitialState())
                    ls.remove()
                    navigate('/')
                }}
                className="w-full bg-blue-500 py-4 text-center text-2xl text-white"
            >
                Sign out
            </Button>
        </div>
    )
}
