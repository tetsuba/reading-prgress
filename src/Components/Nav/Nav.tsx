import NavLinksRightDesktop from "./NavLinksRightDesktop";
import NavLinksLeftDesktop from "./NavLinksLeftDesktop";

export default function Nav() {
    return (
        <nav className="bg-gray-800">
            <div className="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8">
                <div className="flex h-16 items-center justify-between">
                    <NavLinksLeftDesktop />
                    <NavLinksRightDesktop />
                    {/*<NavMobil />*/}
                </div>
            </div>
        </nav>
    )
}