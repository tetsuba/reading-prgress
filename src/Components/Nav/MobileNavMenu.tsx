import { useState } from 'react'

import Button from '../Button/Button'
import Modal from '../Modal/Modal'
import MobileMenu from './MobileMenu'
import Svg from '../Svg/Svg'

export default function MobileNavMenu() {
    const [showMenu, setShowMenu] = useState(false)
    return (
        <div role="dialog" aria-modal="true" className="h-6 md:hidden">
            <Button
                data-testid="mobile-menu-button"
                template="svg"
                className=""
                onClick={() => setShowMenu(true)}
            >
                <Svg icon="menu" />
            </Button>
            {showMenu && (
                <Modal className="h-full">
                    <MobileMenu
                        closeMenu={() => {
                            setShowMenu(false)
                        }}
                    />
                </Modal>
            )}
        </div>
    )
}
