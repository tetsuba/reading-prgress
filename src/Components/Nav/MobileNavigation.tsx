import { useState } from 'react'

import Modal from '../Modal/Modal'
import MobileMenu from '../Modal/MobileMenu'
import Display from '../Dispay/Display'
import Button from '../Button/Button'

export default function MobileNavigation() {
    const [showMenu, setShowMenu] = useState(false)
    return (
        <div role="dialog" aria-modal="true" className="h-6 md:hidden">
            <Button
                data-testid="mobile-menu-button"
                icon="menu"
                template="icon"
                onClick={() => setShowMenu(true)}
            />
            <Display value={showMenu}>
                <Modal className="h-full">
                    <MobileMenu
                        closeMenu={() => {
                            setShowMenu(false)
                        }}
                    />
                </Modal>
            </Display>
        </div>
    )
}
