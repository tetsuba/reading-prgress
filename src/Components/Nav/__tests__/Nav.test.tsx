import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import {
    WrapperWith_Store_Router,
    WrapperWith_Store_Query_Router
} from '../../../vitest-setup'
import Nav from '../Nav'
import store from '../../../store/store'
import { updateUser } from '../../../store/user/userSlice'

const mockNavigate = vi.fn()
vi.mock('react-router-dom', async () => {
    const actual = await vi.importActual('react-router-dom')
    return {
        ...(actual as object),
        useNavigate: () => mockNavigate
    }
})

const mockUser = {
    data: {
        firstName: 'Bob',
        lastName: 'Bob',
        email: 'Bob@Bob.com',
        id: 1
    },
    token: 'MockToken'
}

describe('Nav', () => {
    describe('not authenticated', () => {
        test('clicking on the login button', async () => {
            const { asFragment } = render(
                <WrapperWith_Store_Query_Router pathname="/">
                    <Nav />
                </WrapperWith_Store_Query_Router>
            )
            expect(asFragment()).toMatchSnapshot()
            fireEvent.click(screen.getByText('Log in'))
            await waitFor(() => expect(screen.getByTestId('login-view')))
            fireEvent.click(screen.getByTestId('modal-close'))
            expect(screen.queryByTestId('login-view')).toBeNull()
        })
    })
    describe('authenticated', () => {
        test('clicking on the sign out button [desktop]', async () => {
            store.dispatch(updateUser(mockUser))

            render(
                <WrapperWith_Store_Router pathname="/profile">
                    <Nav />
                </WrapperWith_Store_Router>
            )

            fireEvent.click(screen.getByTestId('user-menu-button'))
            fireEvent.click(screen.getByText('Sign out'))
            expect(screen.queryByTestId('user-menu')).toBeNull()
            await waitFor(() => expect(mockNavigate).toHaveBeenCalled())
        })
        test('clicking on the sign out button [mobile]', async () => {
            store.dispatch(updateUser(mockUser))
            render(
                <WrapperWith_Store_Router pathname="/profile">
                    <Nav />
                </WrapperWith_Store_Router>
            )
            fireEvent.click(screen.getByTestId('mobile-menu-button'))
            fireEvent.click(screen.getByTestId('sign-out-mobile-menu'))
            await waitFor(() => expect(mockNavigate).toHaveBeenCalled())
        })
        test('clicking on a mobile navigation menu link', async () => {
            store.dispatch(updateUser(mockUser))
            render(
                <WrapperWith_Store_Router pathname="/profile">
                    <Nav />
                </WrapperWith_Store_Router>
            )
            fireEvent.click(screen.getByTestId('mobile-menu-button'))
            expect(screen.getByTestId('mobile-nav-menu'))
            fireEvent.click(screen.getAllByTestId('mobile-menu-link')[0])
            expect(screen.queryByTestId('mobile-nav-menu')).toBeNull()
            await waitFor(() => expect(mockNavigate).toHaveBeenCalled())
        })
    })
})
