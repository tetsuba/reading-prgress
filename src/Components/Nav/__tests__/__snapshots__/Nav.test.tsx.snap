// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Nav > not authenticated > clicking on the login button 1`] = `
<DocumentFragment>
  <div
    class="h-16 md:h-auto"
  >
    <nav
      class="fixed z-10 w-full bg-gray-800 md:relative"
    >
      <div
        class="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8"
      >
        <div
          class="flex h-16 justify-between"
        >
          <div
            class="flex items-center"
          >
            <svg
              class="h-6 w-6 stroke-white"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 6.042A8.967 8.967 0 006 3.75c-1.052 0-2.062.18-3 .512v14.25A8.987 8.987 0 016 18c2.305 0 4.408.867 6 2.292m0-14.25a8.966 8.966 0 016-2.292c1.052 0 2.062.18 3 .512v14.25A8.987 8.987 0 0018 18a8.967 8.967 0 00-6 2.292m0-14.25v14.25"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <div
              class="hidden md:block"
            >
              <div
                class="ml-10 flex items-baseline space-x-4"
              >
                <a
                  aria-current="page"
                  class="text-sm font-medium rounded-md px-3 py-2  bg-gray-900 text-white"
                  href="/"
                >
                  Home
                </a>
                <a
                  class="text-sm font-medium rounded-md px-3 py-2 text-gray-300 hover:bg-gray-700 hover:text-white"
                  href="/register"
                >
                  Register
                </a>
              </div>
            </div>
          </div>
          <div
            class="flex items-center"
          >
            <button
              class="rounded-lg font-medium text-base text-center px-5 py-2.5 text-gray bg-gray-200 hover:bg-gray-300 "
              template="secondary"
            >
              Log in
            </button>
          </div>
        </div>
      </div>
    </nav>
  </div>
</DocumentFragment>
`;
